File: src/startpage.cpp
#include <QByteArray>
#include <QDebug>
#include <QDesktopServices>
#include <QDir>
#include <QList>
#include <QListWidget>
#include <QListWidgetItem>
#include <QMenu>
#include <QProcess>
#include <QSettings>
#include <QStringList>
#include <QTextEdit>

#include "startpage.h"
#include "ui_startpage.h"
#include "mainwindow.h"
#include "qsiproject.h"
#include "util.h"

StartPage::StartPage(QWidget *parent) : QWidget(parent), ui(new Ui::StartPage) {
	ui->setupUi(this);
	this->baseInfoHTML = ui->gameInfoText->toHtml();
	QList<int> mainSplitterList;
	mainSplitterList << 250 << 100;
	ui->splitter->setSizes(mainSplitterList);
	this->setGameInfoText("","","","","");
	this->gameList = QList<QSIProject*>();
	ui->projectIcons->setContextMenuPolicy(Qt::CustomContextMenu);
	this->rightClickMenu = new QMenu(ui->projectIcons);
	this->startGameAction = this->rightClickMenu->addAction("Start game");
	this->loadProjectAction = this->rightClickMenu->addAction("Load project");
	this->openProjectDirAction = this->rightClickMenu->addAction("Open project directory");
	this->rightClickMenu->addSeparator();
	this->rightClickMenu->addAction("Refresh game list");

	this->refreshGameList();
}

StartPage::~StartPage() {
	delete ui;
}

QString StartPage::getGameInfoText() {
	return ui->gameInfoText->toHtml();
}

void StartPage::refreshGameList() {
	this->gameList.clear();
	ui->projectIcons->clear();
	QStringList projectDirs;
	QSettings settings;
	int numProjectDirs = settings.beginReadArray("projectDirs");

	for(int p = 0; p < numProjectDirs; ++p) {
		settings.setArrayIndex(p);
		QDir dir(settings.value("directory").toString());

		if(!dir.exists() || dir.isEmpty()) return;
		QFileInfoList dirs = dir.entryInfoList(QDir::NoDotAndDotDot|QDir::Dirs);

		int numDirs = dirs.length();
		for(int d = 0; d < numDirs; d++) {
			QSIProject* qp = new QSIProject(dirs.at(d).absoluteFilePath());
			if(qp->name == "") {
				qp->name = dirs.at(d).fileName();
				qp->width = -1;
				qp->height = -1;
			}
			this->gameList << qp;
		}
	}
	settings.endArray();
	int listSize = this->gameList.length();
	for(int m = 0; m < listSize; m++) {
		QListWidgetItem* item = new QListWidgetItem(QIcon(":/icons/sphere-icon.png"), this->gameList.at(m)->name);
		item->setTextAlignment(Qt::AlignCenter);
		item->setSizeHint(QSize(150,80));
		ui->projectIcons->addItem(item);
	}
}

void StartPage::setGameInfoText(QString name, QString author, QString resolution, QString path, QString description) {
	ui->gameInfoText->setHtml(this->baseInfoHTML.right(this->baseInfoHTML.length())
		.replace("[PROJECTNAME]", name)
		.replace("[PROJECTAUTHOR]", author)
		.replace("[PROJECTRESOLUTION]", resolution)
		.replace("[PROJECTPATH]", path)
		.replace("[PROJECTDESCRIPTION]", description)
	);
}

void StartPage::on_projectIcons_customContextMenuRequested(const QPoint &pos) {
	QListWidgetItem* selected = ui->projectIcons->itemAt(pos);

	bool actionsEnabled = (selected != nullptr);
	this->startGameAction->setEnabled(false);
	this->loadProjectAction->setEnabled(actionsEnabled);
	this->openProjectDirAction->setEnabled(actionsEnabled);

	QAction* choice = this->rightClickMenu->exec(ui->projectIcons->mapToGlobal(pos));

	if(!choice || !this->currentProject) return;
	QString text = choice->text();
	if(text == "Start game") {

	} else if(text == "Load project")  {
		MainWindow::instance()->openProject(this->gameList.at(ui->projectIcons->currentRow())->getPath(true));
	} else if(text == "Open project directory") {
		QDesktopServices::openUrl(QUrl(this->currentProject->getPath(false)).path());
	} else if(text == "Refresh game list") {
		this->refreshGameList();
	}
}

void StartPage::on_projectIcons_itemActivated(QListWidgetItem *item) {
	MainWindow::instance()->openProject(this->gameList.at(ui->projectIcons->currentRow())->getPath(true));
}

void StartPage::on_projectIcons_itemSelectionChanged() {
	int currentRow = ui->projectIcons->currentRow();
	if(this->gameList.length() == 0) return;
	this->currentProject = this->gameList.at(currentRow);
	if(this->currentProject == nullptr) return;

	QString displayResolution = "";
	if(this->currentProject->width > 0 && this->currentProject->height > 0) {
		displayResolution = QString::number(this->currentProject->width) + "x" + QString::number(currentProject->height);
	}
	this->setGameInfoText(
		this->currentProject->name,
		this->currentProject->author,
		displayResolution,
		this->currentProject->getPath(false),
		this->currentProject->summary
	);
}

File: widgets/startpage.py
from PySide6.QtCore import Qt, Signal, Slot, QPoint, QSettings, QDir, QSize
from PySide6.QtGui import QAction
from PySide6.QtWidgets import QWidget, QMenu, QListWidgetItem

from qsiproject import QSIProject

from ui.ui_startpage import Ui_StartPage

class StartPage(QWidget):
	ui: Ui_StartPage
	printWarnings: bool
	baseInfoHTML: str
	gameList: list[QSIProject]
	currentProject: QSIProject
	rightClickMenu: QMenu
	startGameAction: QAction
	loadProjectAction: QAction
	openProjectDirAction: QAction
	refreshGameListAction: QAction
	projectLoaded: Signal = Signal(QSIProject)

	def __init__(self, parent:QWidget = None, windowType:Qt.WindowType = Qt.WindowType.Widget, printWarnings:bool = False):
		super().__init__(parent, windowType)
		self.ui = Ui_StartPage()
		self.ui.setupUi(self)
		self.printWarnings = printWarnings
		self.baseInfoHTML = self.ui.gameInfoText.toHtml()
		self.ui.splitter.setSizes((250, 100))
		self._setGameInfoText("","","","","")
		self.gameList = []
		self.rightClickMenu = QMenu(self.ui.projectIcons)

		self.currentProject = None
		self.startGameAction = self.rightClickMenu.addAction("Start game")
		self.loadProjectAction = self.rightClickMenu.addAction("Load project")
		self.openProjectDirAction = self.rightClickMenu.addAction("Open project directory")
		self.rightClickMenu.addSeparator()
		self.refreshGameListAction = self.rightClickMenu.addAction("Refresh game list")
		self.refreshGameList()

		self.ui.projectIcons.customContextMenuRequested.connect(self.onProjectIconsCustomContextMenuRequested)
		self.ui.projectIcons.itemActivated.connect(self.onProjectIconsItemActivated)
		self.ui.projectIcons.currentItemChanged.connect(self.onProjectIconsCurrentItemChanged)


	def selectedGame(self) -> QSIProject|None:
		selectedIndexes = self.ui.projectIcons.selectedIndexes()
		if len(selectedIndexes) == 0 or selectedIndexes[0].row() >= len(self.gameList):
			return None
		return self.gameList[selectedIndexes[0].row()]


	@Slot(QPoint)
	def onProjectIconsCustomContextMenuRequested(self, pos:QPoint):
		selected = self.ui.projectIcons.itemAt(pos)
		actionsEnabled = selected is not None
		self.startGameAction.setEnabled(actionsEnabled)
		self.loadProjectAction.setEnabled(actionsEnabled)
		self.openProjectDirAction.setEnabled(actionsEnabled)
		choice = self.rightClickMenu.exec(self.ui.projectIcons.mapToGlobal(pos))
		if choice is None:
			return
		match choice:
			case self.loadProjectAction:
				self.ui.projectIcons.itemActivated.emit(selected)


	@Slot(QListWidgetItem)
	def onProjectIconsItemActivated(self, item: QListWidgetItem):
		selectedIndex = self.ui.projectIcons.indexFromItem(item).row()
		self.currentProject = self.gameList[selectedIndex]
		self._setGameInfoText(
			self.currentProject.name,
			self.currentProject.author,
			self.currentProject.getResolutionString(),
			self.currentProject.projectFilePath,
			self.currentProject.summary)
		self.projectLoaded.emit(self.currentProject)

	@Slot(QListWidgetItem, QListWidgetItem)
	def onProjectIconsCurrentItemChanged(self, cur: QListWidgetItem, prev: QListWidgetItem):
		pass

	def getGameInfoText(self) -> str:
		self.ui.gameInfoText.toHtml()

	def refreshGameList(self):
		self.gameList.clear()
		settings = QSettings()
		numProjectDirs = settings.beginReadArray("projectDirs")
		for p in range(numProjectDirs):
			settings.setArrayIndex(p)
			dir = QDir(settings.value("directory"))
			if not dir.exists() or dir.isEmpty():
				continue
			dirs = dir.entryInfoList()
			for projDir in dirs:
				fn = projDir.fileName()
				if fn == "." or fn == "..":
					continue
				project = QSIProject()
				if project.open(projDir.filePath(), self.printWarnings):
					self.gameList.append(project)
		settings.endArray()

		self.gameList.sort(key=lambda game: game.name.lower())
		self.ui.projectIcons.clear()
		for project in self.gameList:
			item = QListWidgetItem(project.getIcon(), project.name)
			item.setTextAlignment(Qt.AlignmentFlag.AlignCenter)
			self.ui.projectIcons.addItem(item)


	def _setGameInfoText(self, name:str, author:str, resolution:str, path:str, description:str):
		self.ui.gameInfoText.setHtml(self.baseInfoHTML
			.replace("[PROJECTNAME]", name)
			.replace("[PROJECTAUTHOR]", author)
			.replace("[PROJECTRESOLUTION]", resolution)
			.replace("[PROJECTPATH]", path)
			.replace("[PROJECTDESCRIPTION]", description)
		)

